unstoppable-ios-app/domains-manager-ios/Modules/Home/SendCryptoAsset/CryptoSender/CryptoSender+Entities.swift

```diff
@@ -80,25 +110,93 @@ struct CryptoSendingSpec {
+        
+        switch token {
+        case .eth, .matic: self.amount = EVMCoinAmount(units: units)
+        default: self.amount = ERC20Token(units: units, decimals: try token.getContractDecimals(for: .Ethereum))
+        }
+        
         self.token = token
-        self.amount = amount
         self.speed = speed
     }
 }
 
 extension CryptoSender {
     enum Error: Swift.Error {
         case sendingNotSupported
+        case tokenNotSupportedOnChain
+        case decimalsNotIdentified
         case failedFetchGasPrice
+        case failedCreateSendTransaction
         case insufficientFunds
+        case invalidAddresses
     }
     
     enum SupportedToken: String {
         case eth = "ETH"
         case matic = "MATIC"
+        case usdt = "USDT"
+        case usdc = "USDC"
+        case bnb = "BNB"
+        case weth = "WETH"
+        
+        static let array: [CryptoSender.SupportedToken :
+                            [BlockchainType : (mainnet: String,
+                                               testnet: String?,
+                                               decimals: UInt8)]] =
+        [
+                .usdt: [.Ethereum: (mainnet: "0xdAC17F958D2ee523a2206206994597C13D831ec7",
+                                testnet: "0xaA8E23Fb1079EA71e0a56F48a2aA51851D8433D0", // sepolia
+                                decimals: 6)],
+            
+                .usdt: [.Matic: (mainnet: "0xc2132D05D31c914a87C6611C10748AEb04B58e8F",
+                                 testnet: nil, // amoy
+                                 decimals: 6)],
+            
+                .bnb: [.Ethereum: (mainnet: "0xB8c77482e45F1F44dE1745F52C74426C631bDD52",
+                                   testnet: nil, // sepolia
+                                   decimals: 18)],
+            
+                .bnb: [.Matic: (mainnet: "0x3BA4c387f786bFEE076A58914F5Bd38d668B42c3",
+                                testnet: nil, // amoy
+                                decimals: 18)],
+            
+                .usdc: [.Ethereum: (mainnet: "0xA0b86991c6218b36c1d19D4a2e9Eb0cE3606eB48",
+                                    testnet: "0x1c7D4B196Cb0C7B01d743Fbc6116a902379C7238", // sepolia
+                                    decimals: 6)],
+            
+                .usdc: [.Matic: (mainnet: "0x3c499c542cEF5E3811e1192ce70d8cC03d5c3359",
+                                 testnet: nil, // amoy
+                                 decimals: 6)],
+            
+                .weth: [.Ethereum: (mainnet: "0xC02aaA39b223FE8D0A0e5C4F27eAD9083C756Cc2",
+                                    testnet: "0x7b79995e5f793A07Bc00c21412e50Ecae098E7f9", // sepolia
+                                    decimals: 18)],
+            
+                .weth: [.Matic: (mainnet: "0x7ceB23fD6bC0adD59E62ac25578270cFf1b9f619",
+                                 testnet: nil, // amoy
+                                 decimals: 18)],
```

Comment:
Is it expected for some testnet values to be nil?

Reply:
@Oleg-Pecheneg Unfortunately, not every token has an established smart contract on testnets. Some testnets have a fake contract, some have several and we cannot detect which is legit. Anyways this info will be updated when needed.